src/lib/badge/badge.ts

```diff
@@ -0,0 +1,221 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Directive, Input, Renderer2, ElementRef} from '@angular/core';
+import {MatIconRegistry} from '@angular/material/icon';
+import {coerceBooleanProperty} from '@angular/cdk/coercion';
+import {ThemePalette} from '@angular/material/core';
+
+let nextId = 0;
+
+ /** @docs-private */
+export class MatBadgeBase {
+
+  /** Theme for the badge */
+  get matBadgeColor(): ThemePalette { return this._color; }
+  set matBadgeColor(value: ThemePalette) {
+    const colorPalette = value;
+
+    if (colorPalette !== this._color) {
+      if (this._color) {
+        this._renderer.removeClass(this._elementRef.nativeElement, `mat-badge-${this._color}`);
+      }
+      if (colorPalette) {
+        this._renderer.addClass(this._elementRef.nativeElement, `mat-badge-${colorPalette}`);
+      }
+
+      this._color = colorPalette;
+    }
+  }
+  private _color: ThemePalette = 'primary';
+
+  /** Whether the badge should overlap its contents or not */
+  set matBadgeOverlap(val: boolean) {
+    this._overlap = coerceBooleanProperty(val);
+  }
+  get matBadgeOverlap(): boolean {
+    return this._overlap;
+  }
+  private _overlap: boolean = true;
+
+  /** Position the badge should reside; 'above|below before|after' */
+  set matBadgePosition(val: string) {
+    this._position = val;
+    this._isAbove = val.indexOf('below') === -1;
+    this._isAfter = val.indexOf('before') === -1;
+  }
+  get matBadgePosition(): string {
+    return this._position;
+  }
+  private _position: string = 'above after';
+
+  _id: number = nextId++;
+  _isAbove: boolean = true;
+  _isAfter: boolean = true;
+
+  constructor(public _renderer: Renderer2, public _elementRef: ElementRef) {}
+
+  /** Clears the previous badge content */
+  _clearContents(): void {
+    const contents = document.getElementById(`badge-content-${this._id}`);
+    if (contents) {
+      this._renderer.removeChild(this._elementRef.nativeElement, contents);
+    }
+  }
+}
+
+/** Directive to display a text badge. */
+@Directive({
+  selector: '[matBadge]',
+  inputs: ['matBadgeColor', 'matBadgeOverlap', 'matBadgePosition'],
+  host: {
+    'class': 'mat-badge',
+    '[class.mat-badge-overlap]': '_overlap',
+    '[class.mat-badge-above]': '_isAbove',
+    '[class.mat-badge-below]': '!_isAbove',
+    '[class.mat-badge-before]': '!_isAfter',
+    '[class.mat-badge-after]': '_isAfter',
+  },
+})
+export class MatBadge extends MatBadgeBase {
+
+  @Input('matBadge')
+  set content(val: string) {
+    this._content = val;
+    this._clearContents();
+    this._setContent();
+  }
+  get content(): string { return this._content; }
+  private _content: string;
+
+  constructor(_renderer: Renderer2, _elementRef: ElementRef) {
+    super(_renderer, _elementRef);
+  }
+
+  /** Injects a span element into the DOM with the content. */
+  private _setContent(): void {
+    const content = document.createElement('span');
```

Comment:
I don't see why this has to create and insert elements manually. Could you wrap the element via content projection?

Reply:
As far as I know, directives don't have content projection. You can't have multiple components on the same element either so that would't work. The SVG icons do a similar insertion and removal so this isn't the first place we are doing this type of thing.

I had this as a component at first but that API wasn't what @jelbourn had designed.